rfcs/text/0032-structured-benchmark-programs.md

```diff
@@ -0,0 +1,157 @@
+- *Quantum Metropolis Sampling*: [Temme et al., 2011](https://arxiv.org/abs/0911.3635)
+- *ML-QAE*: [Suzuki et al., 2020](https://arxiv.org/abs/1904.10246)
+
+## Conditionals
+
+*This class of benchmark programs utilizes conditionals (such as `if`) to implement protocols or algorithms.*
+
+- *Quantum Teleportation*: Quantum teleportation is a canonical primitive for quantum communication and distributed computing. It utilizes conditionals based on measurement results to apply the appropriate correction operations. ([Bennett et al., 1993](https://doi.org/10.1103/PhysRevLett.70.1895))
+- *Block Encoding*: Block encoding embeds classical data or operators into quantum states. It underpins advanced algorithms like QSVT. It utilizes conditionals to apply different operations. ([Quantum Singular Value Transformation (QSVT)](https://arxiv.org/abs/1806.01838), [Low & Chuang, 2016](https://arxiv.org/abs/1606.02685))
+
+## Mixed Control Flow
+
+*This class of benchmark programs combines loops and conditionals to implement more complex algorithms or protocols.*
+
+- *Quantum Multiplexers*: Quantum multiplexers apply different unitary operations based on the state of control qubits, effectively implementing conditionals in quantum circuits. ([Quantum multiplexers in circuit synthesis](https://arxiv.org/abs/quant-ph/0410066))
+
+## Dynamic Qubit (Re-)Allocation
+
+*This class of benchmark programs allocates qubits dynamically during execution, based on runtime conditions or within loop bodies.*
+
+- *Iterative Quantum Fourier Transform (iQFT)*: Rather than allocating all qubits upfront, just a single qubit is allocated at the start of execution and then reset during each loop iteration of QFT.
+- *Iterative Quantum Phase Estimation (iQPE)*: Similar to iQFT, iQPE allocates a single qubit (in addition to the phase qubit "psi") and reuses it across iterations to estimate the phase of a unitary operator.
+- *Shor's Algorithm*: The primitives used in Shor's algorithm can similarly be implemented using dynamic qubit reallocation to minimize the total number of qubits required.
+
+## Loosely Coupled Hybrid Programs
+
+*Variational and hybrid algorithms naturally combine quantum circuits with classical optimization loops. They are loosely coupled because the quantum and classical parts interact only at well-defined synchronization points.*
+
+- *VQE*: A variational quantum eigensolver algorithm that uses a classical optimizer to minimize the energy of a parameterized quantum circuit which implements several iterations of some ansatz. ([Peruzzo et al., 2014](https://arxiv.org/abs/1304.3061))
+- *QAOA*: A quantum approximate optimization algorithm that uses a classical optimizer to find optimal parameters for a parameterized quantum circuit. It iteratively applies a "problem" and "mixer" operator using loop structures. ([Farhi et al., 2014](https://arxiv.org/abs/1411.4028))
```

Comment:
We might want to explicitly declare this as out-of-scope for this RFC.
The main difference here is that the quantum-classical interaction does _not_ have to happen within the coherence time of the qubits.
This might be an important distinction to make more broadly when we refer to structured programs onwards.